src/FlightController.cpp

```diff
@@ -4,106 +4,293 @@
+        for(uint8_t i(0); i<msp::msg::MAX_MAPPABLE_RX_INPUTS; ++i) {
+            channel_map_[i] = i;
         }
     }
     else {
         // get channel mapping from MSP_RX_MAP
-        msp::msg::RxMap rx_map;
-        client.request(rx_map);
-        channel_map = rx_map.map;
+        msp::msg::RxMap rx_map(fw_variant_);
+        client_.sendMessage(rx_map);
+        std::cout << rx_map;
+        channel_map_ = rx_map.map;
+    }
+    
+    return true;
+}
+
+bool FlightController::disconnect(bool blocking, double timeout)
+{
+    stopMspControl();
+    return client_.disconnect();
+
+}    
+    
+void FlightController::arm(bool blocking) {
+    armed_ = true;
+}
+
+void FlightController::disarm(bool blocking) {
+    armed_ = false;
+}
+
+bool FlightController::armSet() {
+    return armed_;
+}
+
+
+void FlightController::printDebug(bool on)
+{
+    client_.printDebug(on);
+}
+
+
+void FlightController::setFlightMode(FlightMode mode)
+{
+    {
+        std::lock_guard<std::mutex> lock(msp_updates_mutex);
+        flight_mode_ = mode;
+    }
+    generateMSP();
+}
+
+FlightMode FlightController::getFlightMode()
+{
+    return flight_mode_;
+}
+
+void FlightController::setControlSource(ControlSource source)
+{
+    if (source == control_source_) return;
+    //client_.printDebug(true);
+    
+    msp::msg::RxConfig rxConfig(fw_variant_);
+    if (!client_.sendMessage(rxConfig)) std::cout << "client_.sendMessage(rxConfig) failed" << std::endl;
+    std::cout << rxConfig;
+    
+    msp::msg::SetRxConfig setRxConfig(fw_variant_);
+    static_cast<msp::msg::RxConfigSettings&>(setRxConfig) = static_cast<msp::msg::RxConfigSettings&>(rxConfig);
+    
+    if (source == ControlSource::SBUS) {
+        setRxConfig.receiverType = 3;
+        setRxConfig.serialrx_provider = 2;
+    } else if (source == ControlSource::MSP) {
+        setRxConfig.receiverType = 4;
+    }
+    //std::cout << setRxConfig;
+    client_.sendMessage(setRxConfig);
+    
+    if (source == ControlSource::MSP) startMspControl();
+    else stopMspControl();
+    
+    control_source_ = source;
+    //client_.printDebug(false);
+}
+
+ControlSource FlightController::getControlSource()
+{
+    msp::msg::RxConfig rxConfig(fw_variant_);
+    client_.sendMessage(rxConfig);
+    
+    if (rxConfig.receiverType && rxConfig.receiverType() == 4) return ControlSource::MSP;
+    else if (rxConfig.serialrx_provider() == 2) return ControlSource::SBUS;
+    return ControlSource::OTHER;
+    
+}
+
+void FlightController::setRPYT(std::array<double,4>& rpyt)
+{
+    //std::cout << "void FlightController::setRPYT(std::array<double,4>& rpyt)" << std::endl;
+    {
+        std::lock_guard<std::mutex> lock(msp_updates_mutex);
+        rpyt_.swap(rpyt);
     }
+    generateMSP();
+    //std::cout << "finished setRPYT" << std::endl;
 }
 
-bool FlightController::isFirmware(const FirmwareType firmware_type) {
-    return firmware == firmware_type;
+void FlightController::startMspControl()
```

Comment:
What is this method doing (document)? "Control" sounds like it's providing something autonomous. Select a better name that reflects the function of the method.

Reply:
It actually starts a `PeriodicTimer` in the `FlightController` which sends an msp::msg::Rc message at 10Hz using cached values. This prevents the FC from disarming due to loss of RC input. I could probably call it "enablePeriodicMspControl" or something.

Reply:
I am not sure if we should do this. There are pros and cons to this.
The main issue is, that this will prevent the FC going into failsafe mode if no new RC values are received. E.g. if this is used in a ROS node and you do not send new RC values to the node, you actually want the FC to go into failsafe mode.
The only advantage I see is that sometimes people have issues to arm the FC via the dedicated `arm` method because they do not send new RC values after arming frequently enough and the FC goes into failsafe mode. However, the failsafe mode is exactly for this purpose.

What do you think? Should we remove the `arm` method and remove this periodic sending of cached values? In this case, a user would need to periodically send RC to the FC to make sure that it doesn't go into failsafe.

Reply:
Now that I think about this, I am very strongly in favour of removing this functionality from the library. Something like periodically sending cached RC values should not be done on a low level, because it removes the ability for the FC to react on communication loss.
If a user wants such functionality, it should be implemented on a higher level. E.g. by calling the `setRc` method periodically in his code or, if used as ROS node, periodically sending the RC commands to the node.